### ISSUE TITLE
Is it possible to see whole failing element path inside report?

### ISSUE THREAD

**Issue by zunjooo** [pa11y/pa11y#445]
## Expected behaviour
It would be more helpful to see whole failing element path inside report after test.

## Actual behaviour

I just see part of element's path, for example:
<fieldset> <!-- search-redirect.php is pr...</fieldset>

I am talking about part with "..." after some part of element.

## Steps to reproduce
1. Start test with wikipedia page for example like:
`pa11y http://wikipedia.com`

2. Output looks like this:
 • Error: This fieldset element does not have a name available to an accessibility API. Valid names are: legend element, aria-label undefined, aria-labelledby undefined.
   ├── WCAG2AA.Principle4.Guideline4_1.4_1_2.H91.Fieldset.Name
   ├── #search-form > fieldset
   └── <fieldset> <!-- search-redirect.php is pr...</fieldset>

 • Error: Fieldset does not contain a legend element. All fieldsets should contain a legend element that describes a description of the field group.
   ├── WCAG2AA.Principle1.Guideline1_3.1_3_1.H71.NoLegend
   ├── #search-form > fieldset
   └── <fieldset> <!-- search-redirect.php is pr...</fieldset>

3. Is it possible to see whole path for for example this element?
<fieldset> <!-- search-redirect.php is pr...</fieldset>

Or in case with some of our webpage tests:
<svg class...</a> (select with "html > body > div:nth-child(1) > div > div:nth-child(1) > div > div > div:nth-child(2) > a") - I don't see nothing after "class" part.

## Environment:
Pa11y:      5.1.0
Node.js:    10.1.0
npm:        6.4.1
OS:         18.2.0 (darwin)

**Comment by krukru** [comment]
This issue is related to this snippet 
```js
		function processIssueHtml(element) {
			let outerHTML = null;
			let innerHTML = null;
			if (!element.outerHTML) {
				return outerHTML;
			}
			outerHTML = element.outerHTML;
			if (element.innerHTML.length > 31) {
				innerHTML = `${element.innerHTML.substr(0, 31)}...`;
				outerHTML = outerHTML.replace(element.innerHTML, innerHTML);
			}
			if (outerHTML.length > 251) {
				outerHTML = `${outerHTML.substr(0, 250)}...`;
			}
			return outerHTML;
		}
```
https://github.com/pa11y/pa11y/blob/ddcbb3696a43f775e32399bbafeea7f12c9d6cbb/lib/runner.js#L121-L136

For some reason, long html elements get truncated, which makes finding (and fixing) them very difficult. 
Can you give some context to this rule, we would be happy to make a PR if you think this truncating is no longer needed.

**Comment by krukru** [comment]
We failed to notice that pa11y reporter also gives a css selector, which helps to pinpoint exactly the element in question :)

**Comment by joeyciechanowicz** [comment]
Maybe @hollsk might remember? The only reason I can think of is to keep reports a bit more concise. 

Otherwise I can't see a reason that the `outerHTML` has to be truncated still.

**Comment by hollsk** [comment]
No idea, I'm afraid. @rowanmanning ?

**Comment by rowanmanning** [comment]
Yep, this is because the element's outerHTML and innerHTML are potentially very large, just attributes alone on certain elements are huge (think storing JSON in a `data-` attribute) and sending them between PhantomJS and the Node.js process would time out for many pages back in the day.

Since this we've switched to Chrome/Puppeteer, which may not present the same issue, however  it would still be easy to get out-of-memory errors in Node.js when each of the error objects has an enormous string to describe the exact element's HTML.

This _might_ be acceptable in the Pa11y command-line tool, but it would present a risk for anyone running the webservice and dashboard – you don't want the structure of an HTML page to be able to crash your webservice.

**Comment by joeyciechanowicz** [comment]
> One solution may be to increase the max size of the truncation?

That sounds like a sane idea. Something large enough to generally always print out the whole `outerHHTML`, but not stop things working. Perhaps ~1024 for the `outerHTML`.

Is there a need for a long `innerHTML`?

**Comment by rowanmanning** [comment]
Or another potential solution (though it's complicated): we get the element's `tagName`, loop over its attributes and construct the `outerHTML` ourselves using those. If an individual attribute name or value is longer than we allow then we concat just that attribute, and if the whole thing is longer than ~1024 then we also concat that. We could even prioritise identifying information, e.g. the `id` attribute, putting that first.

**Pros:**
- Keeps the output short
- Prioritises identifying information, possibly making it easier to find elements

**Cons:**
- More code to maintain
- Possibly makes it _harder_ to find elements if people want to search their code for the full snippet

**Comment by joeyciechanowicz** [comment]
I knocked out some code to see what it would look like: https://jsfiddle.net/jz1tLp2d/

Whilest writing it I realised some other edge cases such as attributes with no values, which will clutter up the code a bit more.

I think that simply truncating at `>1024` characters is sufficient for most cases. When someone has blobs of JSON in attributes, or a huge number of attributes, then the information within that 1024 characters should allow them to locate the element. And when not, there is the CSS selector to fall back on.
